File: 16-take-screenshot-on-failure/csharp/Screenshot.cs
using NUnit.Framework;
using NUnit.Framework.Interfaces;
using OpenQA.Selenium;
using OpenQA.Selenium.Firefox;

public class Screenshot
{
    IWebDriver Driver;

    [SetUp]
    public void SetUp()
    {
        Driver = new FirefoxDriver();
    }

    [TearDown]
    public void TearDown()
    {
        if (TestContext.CurrentContext.Result.Outcome.Status.Equals(TestStatus.Failed))
            TakeScreenshot();

        Driver.Quit();
    }

    private void TakeScreenshot()
    {
        string TestName     = TestContext.CurrentContext.Test.FullName;
        string SaveLocation = System.Environment.CurrentDirectory + 
          $"/../../../failshot_{TestName}.png";
        ITakesScreenshot ScreenshotDriver = (ITakesScreenshot) Driver;
        ScreenshotDriver.GetScreenshot().SaveAsFile(SaveLocation);
    }

    [Test]
    public void ScreenShotOnFailure()
    {
        Driver.Navigate().GoToUrl("http://the-internet.herokuapp.com");
        Assert.That(false.Equals(true));
    }
}

File: 16-take-screenshot-on-failure/python/screenshot.py
# -*- coding: utf-8 -*-
"""
Implementation of http://elementalselenium.com/tips/16-take-screenshot-on-failure
"""

import sys
import unittest
from selenium import webdriver


class ScreenShotOnFailure(unittest.TestCase):

    def setUp(self):
        self.driver = webdriver.Firefox()

    def tearDown(self):
        if sys.exc_info()[0]:
            self.driver.save_screenshot("failshot_%s.png" % self._testMethodName)
        self.driver.quit()

    def test_example_1(self):
        driver = self.driver
        driver.get('http://the-internet.herokuapp.com')
        assert driver.title == 'blah blah blah'

if __name__ == "__main__":
    unittest.main()
